### PR TITLE
Tie up loose ends for urclock programmer

### PR DESCRIPTION
**Pull request by stefanrueger**
This PR ties up loose ends for the urclock programmer
 - Add autobaud_sync to avrdude.conf part description
 - Establish avr_*timestamp() library routines for timing
 - Adapt urclock_getsync() to consider legacy bootloaders
 - Update documentation re the urclock programmer
 - Get boot section info from avrdude.conf and not from avrintel.c

In particular, I changed the synchronisation routine to better deal with non-urboot bootloaders. They now work under Linux out of the box, only picoboot needs the new flag `-xstrict`. Urclock and urboot have a fast synchronisation, but other bootloaders *might* need an additional delay of `-xdelay=<n>` ms, depending on the board and the OS (Windows vs Linux). This number <n> can be negative, in which case it reduces the standard delay of 110 ms after board reset.

Under Linux with urclock boards and legacy bootloaders no `-xdelay=...` is needed. The following bash script tests 24 delays from -15 ms to 100 ms.

``` sh
for d in {-15..100..5}; do
  avrdude -curclock -pm328p -xdelay=$d -xshowall >& /dev/null
  echo -n $? | tr 01 .X
done; echo
```
A `.` means the `-xdelay=` value works; an `X` means it fails. Delay 0 (absence of `-xdelay=...`) is the fourth result letter:

| Bootloader | Result | Result using `-xstrict` | Comments |
| --: | --: | --: | :-- |
urboot fixed baud rate|`........................`|`........................`| All good |
urboot autobaud |`........................`|`........................`| All good |
optiboot v8.3 |`X......XXXXXXX.....XXXXX`|`...XXXX........XXXX.....`| Only some `-xdelay=` work |
optiboot v4.4 |`........................`|`XXXXXXXXXXXXXXXXXXXXXXXX`| Must not use `-xstrict` |
picoboot |`XXXXXXXXXXXXXXXXXXXXXXXX`|`........................`| Must use `-xstrict` |

urclock-urboot-test

I also tested synchronisation times with `/bin/time -f%e avrdude -qqpm328p -xshowversion [-xstrict]`:
| Bootloader | Version | Synchronisation time | ... using `-xstrict` |
| --: | :-: | --: | :-: |
|urboot fixed baud rate|`u7.6 w-u-jpr-.`|0.23 s|0.22 s|
|urboot autobaud|`u7.7 weu-jPrac`|0.27 s|0.26 s|
|optiboot v8.3|`o8.3 --s-h-r--`|0.41 s|n/a|
|optiboot v4.4|`o4.4 --s-h-r--`|0.81 s|n/a|
|picoboot|`x0.0 .........`|n/a|0.33 s|

### CODE REVIEW

**Comment by mcuee** [comment]
Yes now `-xstrict` works with picoboot, tested under Linux.
`-xdelay` used to work with picoboot (timing branch) but now it does not seem to work.

```
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_timingv2 -C ./avrdude_timingv2.conf -c urclock
 -P /dev/ttyACM0 -p m328p -xshowall -xbootsize=512 -xdelay=10
 121 ms: enter urclock_getsync()
 131 ms zzz(16 ms)
 158 ms zzz(32 ms)
 214 ms: all good, ready to rock

avrdude_timingv2: AVR device initialized and ready to accept instructions
0 2022-06-10 21.21 Blink.ino.standard.hex 924 store 31298 meta 34 boot 512 x0.0 ......... 
vector 0 (RESET) ATmega328P

cuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_pr1198 -C ./avrdude_pr1198.conf -c urclock
 -P /dev/ttyACM0 -p m328p -xshowall -xbootsize=512 -xdelay=100
avrdude_pr1198 warning: attempt 1 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 2 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 3 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 4 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 5 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 6 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 7 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 8 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 9 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 warning: attempt 10 of 10: not in sync
avrdude_pr1198 error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
avrdude_pr1198 error: unable to open programmer urclock on port /dev/ttyACM0

avrdude_pr1198 done.  Thank you.

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_pr1198 -C ./avrdude_pr1198.conf -c urclock
 -P /dev/ttyACM0 -p m328p -xshowall -xstrict

avrdude_pr1198: AVR device initialized and ready to accept instructions
0 2022-06-10 21.21 Blink.ino.standard.hex 924 store 31298 meta 34 boot 512 x0.0 ......... vector 0 (RESET) ATmega328P
```

**Comment by mcuee** [comment]
The old optiboot 4.4 works fine.
```
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_pr1198 -C ./avrdude_pr1198.conf -c urclock
 -P /dev/ttyACM0 -p m328p -U ./blink_m328p.hex 

avrdude_pr1198: AVR device initialized and ready to accept instructions
avrdude_pr1198: device signature = 0x1e950f (probably m328p)
avrdude_pr1198: Note: flash memory has been specified, an erase cycle will be performed.
                To disable this feature, specify the -D option.
avrdude_pr1198: erasing chip
                delaying chip erase until first -U upload to flash
avrdude_pr1198: reading input file ./blink_m328p.hex for flash
                with 1030 bytes in 1 section within [0, 0x405]
                using 9 pages and 122 pad bytes
avrdude_pr1198: preparing flash input for device bootloader
avrdude_pr1198: writing 1030 bytes flash ...

Writing | ################################################## | 100% 5.49 s 

avrdude_pr1198: 1030 bytes of flash written
avrdude_pr1198: verifying flash memory against ./blink_m328p.hex

Reading | ################################################## | 100% 4.13 s 

avrdude_pr1198: 1030 bytes of flash verified

avrdude_pr1198 done.  Thank you.

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_pr1198 -C ./avrdude_pr1198.conf -c urclock -P /dev/ttyACM0 -p m328p -xshowall

avrdude_pr1198: AVR device initialized and ready to accept instructions
0 2022-11-20 12.58 blink_m328p.hex 1030 store 31199 meta 27 boot 512 o4.4 --s-h-r-- vector 0 
(RESET) ATmega328P


```

**Comment by mcuee** [comment]
Results with optboot 8.0 from @MCUdude 
```
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ cat test_pr1198.sh 
for d in {-15..100..5}; do
  ./avrdude_pr1198 -C ./avrdude_pr1198.conf -curclock -P /dev/ttyACM0 -pm328p -xdelay=$d -xshowall -xbootsize=512 >& /dev/null
  echo -n $? | tr 01 .X
done; echo
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./test_pr1198.sh 
XX......XXXXXX......XXXX

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ cat test_pr1198.sh 
for d in {-15..100..5}; do
  ./avrdude_pr1198 -C ./avrdude_pr1198.conf -curclock -P /dev/ttyACM0 -pm328p -xdelay=$d -xshowall -xstrict -xbootsize=512 >& /dev/null
  echo -n $? | tr 01 .X
done; echo

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./test_pr1198.sh 
....XXXX........XXXX....

```

**Comment by mcuee** [comment]
So my test results are similar. I will carry out some tests under Windows tomorrow.

**Comment by stefanrueger** [comment]
> -xdelay used to work with picoboot (timing branch) but now it does not seem to work.

Yes, this is deliberate. I piggy-backed the "strict" functionality onto `-xdelay=...` in the test branch but decided to give this its own option for clarity, sanity and ease of documentation.

**Comment by stefanrueger** [comment]
> I will carry out some tests under Windows tomorrow.

Brilliant! Could you show the test results for the different bootloaders in a table like I did above? I can tell from how the `.` and `X` line up which default delay is best. Ideally, I find one that is OK for Windows, Linux and a range of boards. The most critical test is that of [optiboot **v8.3**](https://raw.githubusercontent.com/stefanrueger/urboot/main/src/all/optiboot_atmega328.hex) *without `-xstrict`*. This is sensitive to -xdelay. I haven't checked in the source why (and frankly, I don't want to: alone the condition when this bootloader is entered is mindbogglingly hard to follow).

I believe MacOS, FreeBSD, etc will show the same results as Linux, so no need to test. Windows will be (slightly) different as the drain timeout means something different there. Different boards take different time to get out of reset, so that may be (and has been) significant.

**Comment by stefanrueger** [comment]
> The most critical test is that of [optiboot v8.3](https://raw.githubusercontent.com/stefanrueger/urboot/main/src/all/optiboot_atmega328.hex) without -xstrict.

... and it would be good to see just these 24-letter `.X` test results for optiboot 8.3  in a table for different boards/OS. I can then try to find a column that is all `.` to determine default delays.

**Comment by mcuee** [comment]
> Could you show the test results for the different bootloaders in a table like I did above? I can tell from how the . and X line up which default delay is best. Ideally, I find one that is OK for Windows, Linux and a range of boards. The most critical test is that of [optiboot v8.3](https://raw.githubusercontent.com/stefanrueger/urboot/main/src/all/optiboot_atmega328.hex) without -xstrict. This is sensitive to -xdelay.

Yes I will test under Windows and populate the following table using your format. The Linux test results are similar to yours.

| Bootloader | Result | Result using `-xstrict` | Comments |
| --: | --: | --: | :-- |
urboot fixed baud rate|`........................`|`........................`| All good |
urboot autobaud |`........................`|`........................`| All good |
optiboot v8.3 |`X......XXXXXXX.....XXXXX`|`...XXXX........XXXX.....`| Only some `-xdelay=` work |
optiboot v4.4 |`........................`|`XXXXXXXXXXXXXXXXXXXXXXXX`| Must not use `-xstrict` |
picoboot |`XXXXXXXXXXXXXXXXXXXXXXXX`|`........................`| Must use `-xstrict` |

**Comment by mcuee** [comment]
@stefanrueger 

For optiboot 8.3 hex file the script gets random results under MSYS2 for some delay tims. I need to see if it true or not with manual test. It seems to be true.

```
$ cat test_pr1198.sh
for d in {-15..100..5}; do
  ./avrdude -curclock -P COM3 -pm328p -xdelay=$d -xshowall -xbootsize=512 >& /dev/null
  echo -n $? | tr 01 .X
done; echo

$ sh test_pr1198.sh
..XX.XX......XXXXXX...X.

$ sh test_pr1198.sh
X.XXXX.....XX.XXXX..X...

$ sh test_pr1198.sh
..XXXXX......XXXXX.X....

$ sh test_pr1198.sh
.XXXX.XX....XXXXXX.X....

$ sh test_pr1198.sh
XXXXXX.X.....XXXXX.XX...

```

**Comment by mcuee** [comment]
**TODO**
urboot optiboot 8.3 hex file under different OS.

Optiboot 8.3 | Version | Result | Result   using -xstrict | Comments
-- | -- | -- | -- | --
Linux | `o8.3 --s-h-r--` | `XX......XXXXXXX.....XXXX` | `........................`  |  
Windows | `o8.3 --s-h-r--` | random | `........................` |  
macOS | `o8.3 --s-h-r--` |   |   |

**Comment by mcuee** [comment]
@stefanrueger 

I am thinking that this PR is good to go. There is no easy option to optimize for Optiboot 8.3 hex file.

**Comment by mcuee** [comment]
Just to confirm the above results, I test the optiboot 8.3 hex file using a Nano Clone with CH340 (16MHz, 5V) only for Windows.

**To update**
optiboot 8.3 | Version | Without-xstriact | With-xstrict | Comments
-- | -- | -- | -- | --
Linux | `o8.3 --s-h-r--` | `XXX.....XXXXXXX.....XXXX` | `........................`  |  occassional error with -xstrict
Windows | `o8.3 --s-h-r--` | random | `........................` |  
macOS | `o8.3 --s-h-r--` | `....XXXXXXXXX....XXXXXXX` | `........................` |  

Results without -xstrict Windows | Hex file
`XXXXXX.....XXXX.X.X.....XXXXXX.X...XX.X.XX.X...X` | optiboot 8.3 hex file   from urboot
`.XXX..X.....X.XX.XX.....XXXXXXXXXX..XXXXXXXXX...` | optiboot 8.3 hex file   from urboot
`.XXXXX....X...XXX.X......XXXX..X.....XXXXXXX....` | optiboot 8.3 hex file   from urboot
`..XXX......X.XXXXXX.....X.XXXXXX......XXXXXX....` | optiboot 8.3 hex file   from urboot
`.XXXXX.X...XXXXXX...X...XXXX.X.X.....XXXXXXX....` | optiboot 8.3 hex file   from urboot

**Comment by mcuee** [comment]
@stefanrueger 

If I look at the results, there is only one thing to change, that is to recommend `-xstrict` first and than `-xdelay` when something goes wrong.

```
avrdude error: programmer is not responding; try and vary -xdelay=100 and/or -xstrict
```
is better to be changed to 
```
avrdude error: programmer is not responding; try -xstrict and/or vary -xdelay=100
```

Take note even though optiboot 4.4 (very popular out there due to Arduino) does not work with `-xstrict` but it anyway works out of the box with the default option.

**Comment by mcuee** [comment]
@MCUdude 

I am not so sure if you have some time to try a few different hex files from your optiboot_flash repositories for a few other AVR chips. I do not have very good boards to test optiboot/iurboot in reality as of now.

**Comment by stefanrueger** [comment]
@mcuee Our results almost coincide in Linux. The main difference is ov8.3 and -xstrict, which always work your side, but not on my side.

|Optiboot v8.3| option| result |
|--: | --: | :-- |
| smr Linux |         | `X......XXXXXXX.....XXXXX`|
| smr Linux |-xstrict | `...XXXX........XXXX.....`|
|  xc Linux |         | `XXX.....XXXXXXX.....XXXX`|
|  xc Linux |-xstrict | `........................`|

I wonder whether this is b/c my test setup has an application (blink.hex) on the device (read on).

I believe this weird behaviour of seemingly random delays being necessary to sync is special to optiboot v 8.x and has to do with its ambition to [preserve the reset causes in the MCUSR as much as possible](https://github.com/Optiboot/optiboot/issues/140).

When optiboot v8.x sees the STK_LEAVE_PROGMODE or exits otherwise it shortens the WDT to 16 ms. Now, I still haven't got my head completely around this o8.x-specific effect, but here is what I *suspect* that happens: the issue is that an external reset may not necessarily enter the bootloader but can enter the application instead in certain conditions (probably to show the application "true" reset causes). However, the watchdog may still be set from earlier runs and will trigger later in the application, upon which optiboot *then* decides to run the bootloader, by which time AVRDUDE run will have sent its first handshake and tries to receive a character. So some of the get_sync() go wrong in an almost predictable way for some 25 ms windows (16 ms WTO + ~10 ms for the board to get out of reset).
 
I also suspect this optiboot8.3 specific behaviour is the reason why stk500_getsync() has an "Arduino" clause to reset the board again whenever sync didn't work out: https://github.com/avrdudes/avrdude/blob/baa52e3a02b67dbebc01211b284032d11d73feac/src/stk500.c#L98-L105

All this is, in my opinion, dancing around optiboot's decision when to enter the bootloader in the presumed attempt to preserve the reset cause. (@WestfW, I agree with @SpenceKonde trying hard to preserve the reset cause is a bad idea).

`-c urclock` really is for urboot bootloaders in the first place. Optiboot v8.x is probably better served by -c arduino for its special needs, though with the right `-xdelay=` `-c urboot` also happily serves optiboot v8.x. (`-c urclock` could sync the same as `-c arduino` but then autobaud goes out of the window and the startup time is > 500 ms, so I don't want to make `-c urclock` a prisoner of optiboot)

**Comment by mcuee** [comment]
> @mcuee Our results almost coincide in Linux. The main difference is ov8.3 and -xstrict, which always work your side, but not on my side.

Usually I will upload the blink sketch before the test as it is more stringent test, But I may not do that today as the tests are really intensive. I will double check. Take note it does fail occaionally in my Uno clone under Linux.

> -c urclock really is for urboot bootloaders in the first place. Optiboot v8.x is probably better served by -c arduino for its special needs, though with the right -xdelay= -c urboot also happily serves optiboot v8.x. (-c urclock could sync the same as -c arduino but then autobaud goes out of the window and the startup time is > 500 ms, so I don't want to make -c urclock a prisoner of optiboot)

@stefanrueger 

I totally agree with that. It is perfectly fine that `-c urclock` may not work with optiboot 8.x out of the box. That is why I think you can merge this PR as it is and no real needs to optimize for optiboot 8.x.

**Comment by stefanrueger** [comment]
> Lucky numbers seem to be 25, 30

I have now added 27 to the default delay under windows. Thanks for the intell

**Comment by mcuee** [comment]
I will retest with the optiboot 8.3 hex file under Windows and Linux on the Arduino Uno clone with ATmega16U2 USB to Serial (similar to official Uno).

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1198 -C .\avrdude_pr1198.conf -c urclock -P COM3
 -p m328p -xshowall -xstrict

avrdude_pr1198: AVR device initialized and ready to accept instructions
0 0000-00-00 00.00  application 0 store 0 meta 0 boot 512 o8.3 --s-h-r-- vector 0 (RESET) ATmega328P

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1198 -C .\avrdude_pr1198.conf -c urclock
 -P COM3 -p m328p -xstrict -U .\hex\Blink.ino.standard.hex

avrdude_pr1198: AVR device initialized and ready to accept instructions
avrdude_pr1198: device signature = 0x1e950f (probably m328p)
avrdude_pr1198: Note: flash memory has been specified, an erase cycle will be performed.
                To disable this feature, specify the -D option.
avrdude_pr1198: erasing chip
                delaying chip erase until first -U upload to flash
avrdude_pr1198: reading input file .\hex\Blink.ino.standard.hex for flash
                with 924 bytes in 1 section within [0, 0x39b]
                using 8 pages and 100 pad bytes
avrdude_pr1198: preparing flash input for device bootloader
avrdude_pr1198: writing 924 bytes flash ...

Writing | ################################################## | 100% 7.40 s

avrdude_pr1198: 924 bytes of flash written
avrdude_pr1198: verifying flash memory against .\hex\Blink.ino.standard.hex

Reading | ################################################## | 100% 4.96 s

avrdude_pr1198: 924 bytes of flash verified

avrdude_pr1198 done.  Thank you.

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1198v1 -C .\avrdude_pr1198v1.conf -c urclock
 -P COM3 -p m328p -xshowall -xstrict

avrdude_pr1198v1: AVR device initialized and ready to accept instructions
0 2022-06-10 21.21 Blink.ino.standard.hex 924 store 31298 meta 34 boot 512 
o8.3 --s-h-r-- vector 0 (RESET) ATmega328P

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1198v1 -C .\avrdude_pr1198v1.conf -c urclock
 -P COM3 -p m328p -xshowall -xdelay=0

avrdude_pr1198v1: AVR device initialized and ready to accept instructions
0 2022-06-10 21.21 Blink.ino.standard.hex 924 store 31298 meta 34 boot 512 
o8.3 --s-h-r-- vector 0 (RESET) ATmega328P
```

**Comment by mcuee** [comment]
> > Lucky numbers seem to be 25, 30
> 
> I have now added 27 to the default delay under windows. Thanks for the intell

I am sorry now we do not have lucky number any more with the blink sketch.

**Comment by mcuee** [comment]
@stefanrueger 

I am fine with this PR myself. Let's wait for a few days so that others got a chance to carry out reviews or testing.

**Comment by stefanrueger** [comment]
> I am sorry now we do not have lucky number any more with the blink sketch.

The idea is to change default delay so that -xdelay=0 works both in Lunux and Windows.

**Comment by mcuee** [comment]
> > I am sorry now we do not have lucky number any more with the blink sketch.
> 
> The idea is to change default delay so that -xdelay=0 works both in Lunux and Windows.

Today I tried two ATmega168P Nano Clone with MiniCore and indeed we are lucky so that it is just nice (no need to specify `-xdelay`).

```
C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -c urclock -P COM6 -p m168p -xbootsize=512 -xshowall -xdelay=100

avrdude: AVR device initialized and ready to accept instructions
0 2022-11-28 14.17 Blink_atmega168p.hex 1030 store 14810 meta 32 boot 512 
o8.0 -?s-?-r-- vector 0 (RESET) ATmega168P

C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -c urclock -P COM6 -p m168p -xbootsize=512
 -U .\Blink_atmega168p.hex

avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e940b (probably m168p)
avrdude: Note: flash memory has been specified, an erase cycle will be performed.
         To disable this feature, specify the -D option.
avrdude: erasing chip
         delaying chip erase until first -U upload to flash
avrdude: reading input file .\Blink_atmega168p.hex for flash
         with 1030 bytes in 1 section within [0, 0x405]
         using 9 pages and 122 pad bytes
avrdude: preparing flash input for device bootloader
avrdude: writing 1030 bytes flash ...

Writing | ################################################## | 100% 3.41 s

avrdude: 1030 bytes of flash written
avrdude: verifying flash memory against .\Blink_atmega168p.hex

Reading | ################################################## | 100% 2.30 s

avrdude: 1030 bytes of flash verified

avrdude done.  Thank you.

C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -c urclock -P COM6 -p m168p -xbootsize=512 -xshowall

avrdude: AVR device initialized and ready to accept instructions
0 2022-11-28 14.17 Blink_atmega168p.hex 1030 store 14810 meta 32 boot 512 
o8.0 -?s-?-r-- vector 0 (RESET) ATmega168P
```
